include/freerdp/client/camera.h

```diff
@@ -0,0 +1,188 @@
+
+/* 4 Mbps max encoded bitrate seems to produce reasonably
+ * good quality with H264_RATECONTROL_VBR.
+ */
+#define ECAM_H264_ENCODED_BITRATE 4000000
+
+typedef struct s_ICamHal ICamHal;
+
+typedef struct
+{
+	IWTSPlugin iface;
+	IWTSListener* listener;
+	GENERIC_LISTENER_CALLBACK* hlistener;
+
+	/* HAL interface */
+	ICamHal* ihal;
+	char* subsystem;
+
+	BOOL initialized;
+	BOOL attached;
+
+	UINT32 version;
+	wHashTable* devices;
+
+} CameraPlugin;
+
+typedef struct
+{
+	CAM_MEDIA_FORMAT inputFormat;  /* camera side */
+	CAM_MEDIA_FORMAT outputFormat; /* network side */
+
+} CAM_MEDIA_FORMAT_INFO;
+
+typedef struct
+{
+	BOOL streaming;
+	CAM_MEDIA_FORMAT_INFO formats;
+	CAM_MEDIA_TYPE_DESCRIPTION currMediaType;
+
+	GENERIC_CHANNEL_CALLBACK* hSampleReqChannel;
+	INT nSampleCredits;
+	wStream* sampleRespBuffer;
+
+	H264_CONTEXT* h264;
+	struct SwsContext* sws;
+	int srcLineSizes[4];
+
+} CameraDeviceStream;
+
+#define streamInputFormat(stream)  ((stream)->formats.inputFormat)
+#define streamOutputFormat(stream) ((stream)->formats.outputFormat)
+
+typedef struct
+{
+	IWTSListener* listener;
+	GENERIC_LISTENER_CALLBACK* hlistener;
+	CameraPlugin* ecam;
+	ICamHal* ihal; /* HAL interface, same as used by CameraPlugin */
+	char deviceId[32];
+	CameraDeviceStream streams[ECAM_DEVICE_MAX_STREAMS];
+
+} CameraDevice;
+
+/**
+ * Subsystem (Hardware Abstraction Layer, HAL) Interface
+ */
+
+typedef UINT (*ICamHalEnumCallback)(CameraPlugin* ecam, GENERIC_CHANNEL_CALLBACK* hchannel,
+					 const char* deviceId, const char* deviceName);
+
+/* may run in context of different thread */
+typedef UINT (*ICamHalSampleCapturedCallback)(CameraDevice* dev, int streamIndex,
+						const BYTE* sample, size_t size);
+
+struct s_ICamHal
+{
+	UINT (*Enumerate)(ICamHal* ihal, ICamHalEnumCallback callback,
+			CameraPlugin* ecam, GENERIC_CHANNEL_CALLBACK* hchannel);
+	INT16 (*GetMediaTypeDescriptions)(ICamHal* ihal, const char* deviceId, int streamIndex,
+			const CAM_MEDIA_FORMAT_INFO* supportedFormats, size_t nSupportedFormats,
+			CAM_MEDIA_TYPE_DESCRIPTION* mediaTypes, size_t* nMediaTypes);
+	UINT (*StartStream)(ICamHal* ihal, CameraDevice* dev, int streamIndex,
+			const CAM_MEDIA_TYPE_DESCRIPTION* mediaType,
+			ICamHalSampleCapturedCallback callback);
+	UINT (*StopStream)(ICamHal* ihal, const char* deviceId, int streamIndex);
+	UINT (*Free)(ICamHal* hal);
+};
+
+typedef UINT (*PREGISTERCAMERAHAL)(IWTSPlugin* plugin, ICamHal* hal);
+
+typedef struct
+{
+	IWTSPlugin* plugin;
+	PREGISTERCAMERAHAL pRegisterCameraHal;
+	CameraPlugin* ecam;
+	const ADDIN_ARGV* args;
+
+} FREERDP_CAMERA_HAL_ENTRY_POINTS;
+
+typedef FREERDP_CAMERA_HAL_ENTRY_POINTS* PFREERDP_CAMERA_HAL_ENTRY_POINTS;
+
+/* entry point called by addin manager */
+typedef UINT (*PFREERDP_CAMERA_HAL_ENTRY)(PFREERDP_CAMERA_HAL_ENTRY_POINTS pEntryPoints);
+
+/* common functions */
+UINT ecam_channel_send_generic_msg(CameraPlugin* ecam, GENERIC_CHANNEL_CALLBACK* hchannel, CAM_MSG_ID msg);
```

Comment:
`FREERDP_API` to expose this function, otherwise it should not be in this public header (same below)

Reply:
Moved this file, so it's not a public header